cpp/JobSystem.hpp

```diff
@@ -0,0 +1,175 @@
+                {
+                    threads.emplace_back(hsd::thread(&JobSystem::thread_function, this));
+                }
+            }
+
+            ~JobSystem()
+            {
+                running.store(false);
+
+                for(auto& t : threads)
+                {
+                    t.join();
+                }
+            }
+
+            //Returns the id of the caller from the threadpool index [0, max), and returns -1 if the caller isn't in the list
+            i8 getCurrentThread() const
+            {
+                return -1; //hsd::thread::this_thread::id is not implemented as an equivalent to the std version,
+                           //therefore, this currently cannot be implemented
+
+                /*auto& id = hsd::thread::get_id();
+                for(auto i = 0; i < threads.size(); i++)
+                {
+                    if(id == threads.at(i).get_id())
+                    {
+                        return i;
+                    }
+                }
+                return -1;*/
+            }
+
+            static Job createJob(hsd::JobFn jfunct, void** data = nullptr)
+            {
+                return Job{jfunct, data};
+            }
+
+            void scheduleJob(Job& j, hsd::PRIO p = hsd::PRIO::NORM)
+            {
+                switch(p)
+                {
+                    case hsd::PRIO::HIGH:
+                        high_priority.emplace(std::forward<Job>(j));
+                        counter++;
+                        return;
+                    case hsd::PRIO::NORM:
+                        normal_priority.emplace(std::forward<Job>(j));
+                        counter++;
+                        return;
+                    case hsd::PRIO::LOW:
+                        low_priority.emplace(std::forward<Job>(j));
+                        counter++;
+                        return;
+                }
+            }
+
+            void wait(int target = 0, bool useCurrentThread = true)
+            {
+                while(counter.load() > target)
+                {
+                    if(useCurrentThread)
+                    {
+                        Job job;
+                        if(high_priority.try_pop(job))
+                        {
+                            job.task(job);
+                            counter--;
+                        }
+                        else if(normal_priority.try_pop(job))
+                        {
+                            job.task(job);
+                            counter--;
+                        }
+                        else if(low_priority.try_pop(job))
+                        {
+                            job.task(job);
+                            counter--;
+                        }
+
+                    }
+                }
+            }
+        
+        };
+    }
+
+    extern priv::JobSystem JobSys;
+    
+}
+
+#ifdef JOB_SYSTEM_COMP_UNIT
+hsd::priv::JobSystem JobSys;
```

Comment:
There is this thing called `static inline` variable, it's from C++17

Reply:
This is for a global variable. But I will try to implement this.

Reply:
*coughs*